In [4]:
import pandas as pd
from sqlalchemy import create_engine

# Connect to your Postgres database
# Replace 'your_password' with your actual postgres password
engine = create_engine('postgresql+psycopg2://postgres:admin123@localhost:5432/olist_db')

# List of all tables to check
tables = [
    'customers', 'sellers', 'product_category_name_translation',
    'products', 'orders', 'order_items', 'order_payments',
    'order_reviews', 'geolocation'
]

print("Connected successfully. Ready to profile", len(tables), "tables.")

Connected successfully. Ready to profile 9 tables.


In [5]:
def profile_table(table_name, engine):
    df = pd.read_sql(f"SELECT * FROM {table_name}", engine)
    print(f"\n{'='*60}")
    print(f"TABLE: {table_name}")
    print(f"{'='*60}")
    print(f"Rows: {len(df):,} | Columns: {len(df.columns)}")
    print(f"\nNull counts (columns with nulls only):")
    nulls = df.isnull().sum()
    nulls = nulls[nulls > 0]
    if len(nulls) > 0:
        for col, count in nulls.items():
            pct = round(count / len(df) * 100, 2)
            print(f"  {col}: {count:,} ({pct}%)")
    else:
        print("  None")
    print(f"\nDuplicate rows (full row match): {df.duplicated().sum():,}")
    return df

In [6]:
dataframes = {}
for table in tables:
    dataframes[table] = profile_table(table, engine)


TABLE: customers
Rows: 99,441 | Columns: 5

Null counts (columns with nulls only):
  None

Duplicate rows (full row match): 0

TABLE: sellers
Rows: 3,095 | Columns: 4

Null counts (columns with nulls only):
  None

Duplicate rows (full row match): 0

TABLE: product_category_name_translation
Rows: 73 | Columns: 2

Null counts (columns with nulls only):
  None

Duplicate rows (full row match): 0

TABLE: products
Rows: 32,951 | Columns: 9

Null counts (columns with nulls only):
  product_category_name: 610 (1.85%)
  product_name_lenght: 610 (1.85%)
  product_description_lenght: 610 (1.85%)
  product_photos_qty: 610 (1.85%)
  product_weight_g: 2 (0.01%)
  product_length_cm: 2 (0.01%)
  product_height_cm: 2 (0.01%)
  product_width_cm: 2 (0.01%)

Duplicate rows (full row match): 0

TABLE: orders
Rows: 99,441 | Columns: 8

Null counts (columns with nulls only):
  order_approved_at: 160 (0.16%)
  order_delivered_carrier_date: 1,783 (1.79%)
  order_delivered_customer_date: 2,965 (2.98%)

Dupli

In [7]:
import pandas as pd
pd.set_option('display.max_rows', None)

# Re-run and capture full output for tables we haven't fully seen yet
for table in ['products', 'orders', 'order_items', 'order_payments', 'order_reviews']:
    profile_table(table, engine)


TABLE: products
Rows: 32,951 | Columns: 9

Null counts (columns with nulls only):
  product_category_name: 610 (1.85%)
  product_name_lenght: 610 (1.85%)
  product_description_lenght: 610 (1.85%)
  product_photos_qty: 610 (1.85%)
  product_weight_g: 2 (0.01%)
  product_length_cm: 2 (0.01%)
  product_height_cm: 2 (0.01%)
  product_width_cm: 2 (0.01%)

Duplicate rows (full row match): 0

TABLE: orders
Rows: 99,441 | Columns: 8

Null counts (columns with nulls only):
  order_approved_at: 160 (0.16%)
  order_delivered_carrier_date: 1,783 (1.79%)
  order_delivered_customer_date: 2,965 (2.98%)

Duplicate rows (full row match): 0

TABLE: order_items
Rows: 112,650 | Columns: 7

Null counts (columns with nulls only):
  None

Duplicate rows (full row match): 0

TABLE: order_payments
Rows: 103,886 | Columns: 5

Null counts (columns with nulls only):
  None

Duplicate rows (full row match): 0

TABLE: order_reviews
Rows: 99,224 | Columns: 7

Null counts (columns with nulls only):
  review_comment_

In [8]:
for table in ['order_items', 'order_payments']:
    profile_table(table, engine)


TABLE: order_items
Rows: 112,650 | Columns: 7

Null counts (columns with nulls only):
  None

Duplicate rows (full row match): 0

TABLE: order_payments
Rows: 103,886 | Columns: 5

Null counts (columns with nulls only):
  None

Duplicate rows (full row match): 0


In [9]:
# Check for outliers in key numeric fields
print("=== Price outliers (order_items) ===")
price_df = pd.read_sql("SELECT price, freight_value FROM order_items", engine)
print(price_df.describe())

print("\n=== Payment value outliers (order_payments) ===")
payment_df = pd.read_sql("SELECT payment_value, payment_installments FROM order_payments", engine)
print(payment_df.describe())

=== Price outliers (order_items) ===
               price  freight_value
count  112650.000000  112650.000000
mean      120.653739      19.990320
std       183.633928      15.806405
min         0.850000       0.000000
25%        39.900000      13.080000
50%        74.990000      16.260000
75%       134.900000      21.150000
max      6735.000000     409.680000

=== Payment value outliers (order_payments) ===
       payment_value  payment_installments
count  103886.000000         103886.000000
mean      154.100380              2.853349
std       217.494064              2.687051
min         0.000000              0.000000
25%        56.790000              1.000000
50%       100.000000              1.000000
75%       171.837500              4.000000
max     13664.080000             24.000000


In [10]:
zero_payments = pd.read_sql("""
    SELECT * FROM order_payments 
    WHERE payment_value = 0
""", engine)
print(f"Number of zero-value payments: {len(zero_payments)}")
print(zero_payments['payment_type'].value_counts())
zero_payments.head(10)

Number of zero-value payments: 9
payment_type
voucher        6
not_defined    3
Name: count, dtype: int64


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.0
1,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.0
2,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.0
3,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.0
4,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.0
5,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.0
6,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.0
7,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.0
8,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.0


## Data Cleaning Summary

Profiled all 9 tables in the Olist dataset for nulls, duplicates, and outliers.

### Key findings:
- **customers, sellers, order_items, order_payments**: fully clean, no nulls or duplicates.
- **products**: 610 rows (1.85%) missing category and dimension data — likely incomplete catalog entries; excluded from category-based analysis via inner joins in SQL work.
- **orders**: 160 rows (0.16%) missing `order_approved_at` — consistent with orders cancelled before payment approval.
- **order_reviews**: 88.3% missing review titles, 58.7% missing review text — expected behavior (most customers rate without writing a review), not a data quality issue.
- **order_payments**: 9 rows with ₹0 payment_value — 6 are legitimate voucher lines in multi-payment orders, 3 are an ambiguous 'not_defined' payment type inherited from the source data. Retained as-is.
- **geolocation**: high duplicate count (261,831) reflects multiple coordinate entries per zip code prefix — expected structure, not an error.
- **Price/payment outliers**: wide but plausible range (₹0.85–₹6,735 for item price), consistent with a diverse marketplace catalog. No values indicate data corruption.

### Conclusion
No systemic data quality issues found beyond the two already resolved during database import (missing category mappings, encoding mismatch — see `sql/02_data_quality_fixes.sql`). Dataset is clean and reliable for analysis.